### PR TITLE
feat: canonicalize Reshape allowzero=1 to Reshape allowzero=0 for certain cases.

### PR DESCRIPTION
**Pull request by ttjost**
Conditions:
- If no value in the 'shape' input is set to zero.

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps/Canonicalize.td [hunk 1]
```diff
@@ -685,6 +691,16 @@ def RemoveIdentityReshapePattern2:  Pat<
   // Check that val and out have the same static shape.
   [(IsIdentityReshape $out, $val)]>;
 
+def ReplaceReshapeAllowZeroByReshape:  Pat<
+  // Reshape with allowzero == 1
+  (ONNXReshapeOp:$out $val, $shape, $az),
```
Comment (roberteg16) [inline src/Dialect/ONNX/ONNXOps/Canonicalize.td]: Can we change the name from `az` to `allowZero`. It seems to cryptic 😅

File: test/mlir/onnx/onnx_canonicalization.mlir [hunk 2]
```diff
@@ -400,6 +400,53 @@ func.func @test_reshape_no_fusion(%arg0: tensor<1x3x1152x1x1344xf32>) -> (tensor
...
+// -----
+
+func.func @reshape_allowzero_no_conversion(%arg0: tensor<1x2048x1x0xbf16>) -> (tensor<1x1x1x0x0x2048xbf16>) {
+  %0 = onnx.Constant dense<[1, 1, 1, 0, 0, 2048]> : tensor<6xi64> loc(unknown)
+  %1 = "onnx.Reshape"(%arg0, %0) { allowzero = 1 : si64, onnx_node_name = "Reshape"} : (tensor<1x2048x1x0xbf16>, tensor<6xi64>) -> tensor<1x1x1x0x0x2048xbf16>
+  return %1: tensor<1x1x1x0x0x2048xbf16>
+
+// CHECK-LABEL:  func.func @reshape_allowzero_no_conversion
+// CHECK:        "onnx.Reshape"
+// CHECK-SAME: allowzero = 0
```
Comment (roberteg16) [inline test/mlir/onnx/onnx_canonicalization.mlir]: This seems wrong to me. We can only change allowZero to 0 if new shape does not contain zeros right?

What's the new constant value for this reshape?

File: test/mlir/onnx/onnx_canonicalization.mlir [hunk 3]
```diff
@@ -400,6 +400,53 @@ func.func @test_reshape_no_fusion(%arg0: tensor<1x3x1152x1x1344xf32>) -> (tensor
...
+// -----
+
+func.func @reshape_allowzero_to_reshape_unranked_no_conversion(%arg0: tensor<*xbf16>) -> (tensor<*xbf16>) {
+  %0 = onnx.Constant dense<[1, 1, 1, 2048]> : tensor<4xi64> loc(unknown)
+  %1 = "onnx.Reshape"(%arg0, %0) { allowzero = 1 : si64, onnx_node_name = "Reshape"} : (tensor<*xbf16>, tensor<4xi64>) -> tensor<*xbf16>
+  return %1: tensor<*xbf16>
+
+// CHECK-LABEL:  func.func @reshape_allowzero_to_reshape_unranked_no_conversion
+// CHECK:        "onnx.Reshape"
+// CHECK-SAME: allowzero = 1
```
Comment (roberteg16) [inline test/mlir/onnx/onnx_canonicalization.mlir:470]: This one should be set to 0

File: test/mlir/onnx/onnx_canonicalization.mlir (hunk 2 above)
Comment (ttjost) [reply test/mlir/onnx/onnx_canonicalization.mlir]: yes, I copied the wrong example. It was failing test anyway 😄

**Review by flemairen6 — APPROVED** [review]

**Comment by ttjost** [comment]
I re-read the spec and you can actually have `0` values in `shape` as long as the input has no 0.

File: test/mlir/onnx/onnx_canonicalization.mlir [hunk 4]
```diff
@@ -400,6 +400,89 @@ func.func @test_reshape_no_fusion(%arg0: tensor<1x3x1152x1x1344xf32>) -> (tensor
...
+
+func.func @reshape_allowzero_to_reshape_known_input_zero_value(%arg0 : tensor<5x5x1x32xf32>) -> tensor<*xf32> {
+  %0 = "onnx.Constant"() {value = dense<[-1, 0, 2]> : tensor<3xi64> } : () -> tensor<3xi64>
+  %1 = "onnx.Reshape"(%arg0, %0) { allowzero = 1 : si64 } : (tensor<5x5x1x32xf32>, tensor<3xi64>) -> tensor<*xf32>
+  return %1: tensor<*xf32>
+
+// CHECK-LABEL:  func.func @reshape_allowzero_to_reshape_known_input_zero_value
+// CHECK:        "onnx.Reshape"
+// CHECK-SAME: allowzero = 0
+}
```
Comment (roberteg16) [inline test/mlir/onnx/onnx_canonicalization.mlir]: Isn't this one an invalid reshape? 

[Ref](https://onnx.ai/onnx/operators/onnx__Reshape.html#summary)
>If the attribute ‘allowzero’ is set, it is invalid for the specified shape to contain both a zero value and -1, as the value of the dimension corresponding to -1 cannot be determined uniquely.

Comment (ttjost) [reply test/mlir/onnx/onnx_canonicalization.mlir]: yes, I was changing it the moment you wrote.

**Review by roberteg16 — APPROVED** [review]
LGTM

**Pull request merged**